libasynql/src/poggit/libasynql/base/QueryRecvQueue.php

```diff
@@ -50,10 +50,12 @@ public function publishError(int $queryId, SqlError $error) : void{
 	}
 
 	public function fetchResults(&$queryId, &$results) : bool{
-		$row = $this->shift();
-		if(is_string($row)){
-			[$queryId, $results] = unserialize($row, ["allowed_classes" => true]);
-			return true;
+		while($this->count() > 0){
+			$row = $this->shift();
```

Comment:
You can replace this with a `while(($row = $this->shift()) !== null)` or `while(is_string($row = $this->shift()))` to avoid the extra call (and extra lock/unlock).

Reply:
I think you mean `while(($row = $this->shift()) === null)` because `shift()` will return `null` until it synchronizes, and we're waiting until it does

Reply:
No. It only returns null if the buffer is empty (i.e. count == 0).

Reply:
But the problem here is that `Threaded->count()` will tell us that the buffer isn't empty, but when we do `Threaded->shift()` after that it will return `null` until later. If I do `Threaded->shift()` first the problem will be the same because it will return `null` on waking up from the sleeper and checking we got something in the buffer:
https://github.com/poggit/libasynql/blob/4df7dc2b44de0b6f5cb401dee90c8395655b3ebc/libasynql/src/poggit/libasynql/base/QueryRecvQueue.php#L66